File: native-ios/NaviLive/Sources/Services/Navigation/PedestrianCrossingCore.swift
import Foundation

/// Odróżnia przejścia piesze od jawnie rowerowych elementów OSM.
enum PedestrianCrossingCore {
  private static let allowedFootValues: Set<String> = ["yes", "designated", "official", "permissive"]
  private static let deniedValues: Set<String> = ["no", "false", "use_sidepath"]

  static func isPedestrianCrossing(tags: [String: String]) -> Bool {
    guard !tags.isEmpty else { return false }

    func value(_ key: String) -> String? {
      guard let raw = tags[key] else { return nil }
      let normalized = raw.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
      return normalized.isEmpty ? nil : normalized
    }

    let foot = value("foot")
    let crossingFoot = value("crossing:foot")
    if deniedValues.contains(foot ?? "") || deniedValues.contains(crossingFoot ?? "") {
      return false
    }

    let crossing = value("crossing")
    if deniedValues.contains(crossing ?? "") {
      return false
    }

    let explicitPedestrianAccess = allowedFootValues.contains(foot ?? "") ||
      allowedFootValues.contains(crossingFoot ?? "")
    if value("highway") == "cycleway" && !explicitPedestrianAccess {
      return false
    }
    if crossing == "cycleway" && !explicitPedestrianAccess {
      return false
    }

    let crossingCarriageway = value("crossing:carriageway")
    if crossingCarriageway == "cycleway" && !explicitPedestrianAccess {
      return false
    }
    return value("highway") == "crossing" ||
      value("footway") == "crossing" ||
      crossing != nil ||
      (crossingCarriageway != nil && !deniedValues.contains(crossingCarriageway ?? ""))
  }
}

File: android/app/src/main/java/com/navilive/android/data/routing/PedestrianCrossingCore.kt
package com.navilive.android.data.routing

import java.util.Locale

/** Odróżnia przejścia piesze od jawnie rowerowych elementów OSM. */
internal object PedestrianCrossingCore {

    private val allowedFootValues = setOf("yes", "designated", "official", "permissive")
    private val deniedValues = setOf("no", "false", "use_sidepath")

    fun isPedestrianCrossing(tags: Map<String, String>): Boolean {
        if (tags.isEmpty()) return false

        fun value(key: String): String? = tags[key]
            ?.trim()
            ?.lowercase(Locale.ROOT)
            ?.takeIf { it.isNotBlank() }

        val foot = value("foot")
        val crossingFoot = value("crossing:foot")
        if (foot in deniedValues || crossingFoot in deniedValues) return false

        val crossing = value("crossing")
        if (crossing in deniedValues) return false

        val explicitPedestrianAccess = foot in allowedFootValues ||
            crossingFoot in allowedFootValues
        val highway = value("highway")
        if (highway == "cycleway" && !explicitPedestrianAccess) return false
        if (crossing == "cycleway" && !explicitPedestrianAccess) return false

        val crossingCarriageway = value("crossing:carriageway")
        if (crossingCarriageway == "cycleway" && !explicitPedestrianAccess) return false
        return highway == "crossing" ||
            value("footway") == "crossing" ||
            crossing != null ||
            (crossingCarriageway != null && crossingCarriageway !in deniedValues)
    }
}
